### PR TITLE
NdArray refinement, part 2

### PR DESCRIPTION
**Pull request by bkmartinjr**
Additional refinement of code reachable from NdArray.

Changes:
* Implement delete operation and test cases (for all SOMA foundational types)
* Revise the metadata mapping and SOMACollection to be based upon MutableMapping ABC.
* Add metadata unit tests for all foundational SOMA types
* Add additional SOMACollection unit tests.
* Port SOMAExperiment and SOMAMeasurement to the new SOMACollection.  Establish stricter typing for the pre-defined collections therein.  This entailed the creation of a generic SOMACollectionBase, allowing type narrowing for these cases.

### CODE REVIEW

**Review by johnkerl — APPROVED** [review]
LGTM with nits / minor asks :)

File: apis/python/src/tiledbsoma/soma_collection.py [hunk 1]
```diff
@@ -80,81 +70,54 @@ def get(self, member_name: str) -> TileDBObject:
...
-        return self.get(member_name)
-
-    def set(self, member: TileDBObject, *, relative: Optional[bool] = None) -> None:
+    def set(
+        self,
+        member: TileDBObject,
+        *,
+        child_name: Optional[str] = None,
+        relative: Optional[bool] = None,
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_collection.py:97]: I would prefer doing the fluent work in a separate PR to keep these a bit more focused.

File: apis/python/tests/test_soma_metadata.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import math
+
+import pyarrow as pa
+import pytest
+
+import tiledbsoma as soma
+
+""""
+Metadata handling tests for all SOMA Foundational datatypes.
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/tests/test_soma_metadata.py:9]: fixed

File: apis/python/src/tiledbsoma/soma_experiment.py [hunk 3]
```diff
@@ -53,7 +53,15 @@ def create(self) -> None:
         """
         super().create()
 
-    def __getattr__(self, name: str) -> Any:  # TODO: union type
+    @property
+    def obs(self) -> Any:
+        return self["obs"]
+
+    @property
+    def ms(self) -> Any:
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_experiment.py]: great catch. Fixing this found another bug, which I also fixed.

File: apis/python/src/tiledbsoma/soma_experiment.py [hunk 4]
```diff
@@ -53,7 +53,15 @@ def create(self) -> None:
         """
         super().create()
 
-    def __getattr__(self, name: str) -> Any:  # TODO: union type
+    @property
+    def obs(self) -> Any:
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_experiment.py]: change made

File: apis/python/src/tiledbsoma/soma_collection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_collection.py:97]: See #308

File: apis/python/src/tiledbsoma/soma_collection.py [hunk 5]
```diff
@@ -398,3 +380,7 @@ def _show_metadata(self, recursively: bool = True, indent: str = "") -> None:
                         raise Exception(
                             f"Unexpected object_type found: {object_type} at {obj.uri}"
                         )
+
+
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/soma_collection.py:384]: please add a little comment here about this class' value-add

File: apis/python/src/tiledbsoma/soma_measurement.py [hunk 6]
```diff
@@ -76,7 +79,33 @@ def create(self) -> None:
         """
         super().create()
 
-    def __getattr__(self, name: str) -> Any:
+    @property
+    def var(self) -> Union[SOMADataFrame, SOMAIndexedDataFrame]:
+        return cast(Union[SOMADataFrame, SOMAIndexedDataFrame], self["var"])
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/soma_measurement.py:84]: ooh nice TIL `cast` :D

File: apis/python/src/tiledbsoma/soma_sparse_nd_array.py [hunk 7]
```diff
@@ -149,12 +149,12 @@ def is_sparse(self) -> Literal[True]:
...
-    # TODO
-    #    def get_nnz(self) -> wint:
-    #        """
-    #        Return the number of non-zero values in the array
-    #        """
-    #        return 999
+    @property
+    def nnz(self) -> int:
+        """
+        Return the number of stored values in the array
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/soma_sparse_nd_array.py:155]: wouldn't hurt to add a caveat that explicitly stored zeroes count toward the nnz

File: apis/python/src/tiledbsoma/soma_collection.py [hunk 8]
```diff
@@ -11,8 +21,11 @@
 from .tiledb_object import TileDBObject
 from .tiledb_platform_config import TileDBPlatformConfig
 
+# A collection can hold any sub-type of TileDBObject
+CollectionElementType = TypeVar("CollectionElementType", bound=TileDBObject)
 
-class SOMACollection(TileDBObject):
+
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/soma_collection.py:27]: maybe add a quick comment what is the value-add over `SOMACollection`

**Review by johnkerl — APPROVED** [review]
LGTM, nits only! :)

File: apis/python/src/tiledbsoma/tiledb_object.py [hunk 9]
```diff
@@ -60,6 +60,18 @@ def __init__(
...
         self.metadata = SOMAMetadataMapping(self)
 
+    def delete(self) -> None:
+        """
+        Delete the SOMADataFrame specified with the URI.
+
+        TODO: should this raise an error if the object does not exist?
+        """
+        try:
+            tiledb.remove(self._uri)
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/tiledb_object.py:70]: it does.  I have improved the unit test to confirm.

File: apis/python/src/tiledbsoma/soma_collection.py (hunk 8 above)
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_collection.py:27]: there is no longer a SOMACollection other than this.  There is just SOMACollectionBase

This one implements the general ability to store SOMA objects.  The generic allows type constrains, such as we require in SOMAExperiment and SOMAMeasurement

Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_collection.py:27]: comment added

File: apis/python/src/tiledbsoma/soma_collection.py (hunk 5 above)
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_collection.py:384]: done

File: apis/python/src/tiledbsoma/soma_collection.py (hunk 8 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/soma_collection.py:27]: ahhhh ok thanks!

**Pull request merged**